compiler/src/iree/compiler/GlobalOptimization/RaiseSpecialOps.cpp

```diff
@@ -362,7 +261,319 @@ static FailureOr<Operation *> tryRaiseToView(linalg::GenericOp linalgOp,
 }
 
 //===----------------------------------------------------------------------===//
-// Partial negation and inner slice reverse
+// Named GEMM-like extensions
+//===----------------------------------------------------------------------===//
+
+template <typename OpTy>
+class NamedImplicitCastOpConversion : public OpInterfaceRewritePattern<OpTy> {
+public:
+  using OpInterfaceRewritePattern<OpTy>::OpInterfaceRewritePattern;
+  NamedImplicitCastOpConversion(MLIRContext *ctx, PatternBenefit b = 1)
+      : OpInterfaceRewritePattern<OpTy>(ctx, b) {}
+
+  LogicalResult matchAndRewrite(OpTy namedOp,
+                                PatternRewriter &rewriter) const override {
+    if (!IREE::Flow::isNonNullAndOutsideDispatch(namedOp)) {
+      return failure();
+    }
+
+    // Look for a producer of the given operand that does an elementwise extend
+    // and replace the operand with the source of the elementwise producer.
+    // Returns true if the operand was updated to inform the pattern rewriter
+    // of a change.
+    Type outElementType = getElementTypeOrSelf(namedOp->getResultTypes()[0]);
+    bool didChangeOperand = false;
+    {
+      OpBuilder::InsertionGuard guard(rewriter);
+      Block *block = &namedOp->getRegion(0).front();
+      rewriter.setInsertionPointToStart(block);
+      auto replaceOperandWithTypeCast = [&](OpOperand &operand) {
+        // If the op already has implicit casting semantics for this operand,
+        // do not fuse.
+        if (getElementTypeOrSelf(operand.get().getType()) != outElementType) {
+          return false;
+        }
+        auto producer = operand.get().getDefiningOp<linalg::GenericOp>();
+        if (!producer) {
+          return false;
+        }
+        if (!linalg::isElementwise(producer) ||
+            producer.getNumDpsInputs() != 1 || producer.getNumDpsInits() != 1) {
+          return false;
+        }
+
+        if (!llvm::hasSingleElement(
+                producer.getBlock()->without_terminator())) {
+          return false;
+        }
```

Comment:
We have an util which does the similar things. It does not only look at ExtFOp, but also look at all the CastOpInterface ops that extend the bitwidth. The other difference is that it ignore i1 types, because it is quite tricky in i1 cases. Can we use it here?

https://github.com/iree-org/iree/blob/dd7d4a1441b4be21db2b7ba038bf20d74899c13f/compiler/src/iree/compiler/GlobalOptimization/Utils.h#L26-L40

Reply:
Looking at this helper, it looks like it's returning just the cast, while the logic here is also trying to get the generic op producer. We could try to refactor this for reuse it, but in this PR I'm mostly just shuffling code around.

Reply:
Actually I read the code again and realized you're right, this helper fits nicely. I replaced it, thanks for the suggestion!